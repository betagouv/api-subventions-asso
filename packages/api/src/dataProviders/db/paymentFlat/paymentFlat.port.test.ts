import { PAYMENT_FLAT_ENTITY } from "./__fixtures__/paymentFlatEntity.fixture";
import { PAYMENT_FLAT_DBO } from "./__fixtures__/paymentFlatDbo.fixture";
import paymentFlatPort from "./paymentFlat.port";
import { ObjectId } from "mongodb";
import PaymentFlatAdapter from "../../../modules/paymentFlat/paymentFlatAdapter";
import PaymentsFlatAdapter from "./PaymentFlat.adapter";
const mockDeleteMany = jest.fn();
const mockInsertOne = jest.fn();
const mockUpdateOne = jest.fn();

jest.mock("../../../shared/MongoConnection", () => ({
    collection: () => ({
        deleteMany: mockDeleteMany,
        insertOne: mockInsertOne,
        updateOne: mockUpdateOne,
    }),
}));

describe("PaymentFlat Port", () => {
    describe("insertOne()", () => {
        it("should call insertOne with the correct arguments", async () => {
            await paymentFlatPort.insertOne(PAYMENT_FLAT_ENTITY);
            expect(mockInsertOne).toHaveBeenCalledWith({ ...PAYMENT_FLAT_DBO, _id: expect.any(ObjectId) });
        });
    });

    describe("upsertOne()", () => {
        it("should call upsertOne with the correct arguments", async () => {
            await paymentFlatPort.upsertOne(PAYMENT_FLAT_ENTITY);
            const { _id, ...DboWithoutId } = PAYMENT_FLAT_DBO;
            expect(mockUpdateOne).toHaveBeenCalledWith(
                { uniqueId: PAYMENT_FLAT_ENTITY.uniqueId },
                { $set: DboWithoutId },
                { upsert: true },
            );
        });
    });

    describe("deleteAll()", () => {
        it("should call deleteMany", async () => {
            await paymentFlatPort.deleteAll();
            expect(mockDeleteMany).toHaveBeenCalledWith({});
        });
    });

    describe("cursorFindChorusOnly()", () => {
        let mockCursorFind: jest.SpyInstance;
        let mockDboToEntity: jest.SpyInstance;
        beforeAll(() => {
            mockCursorFind = jest.spyOn(paymentFlatPort, "cursorFind").mockImplementation(jest.fn());
            mockDboToEntity = jest.spyOn(PaymentsFlatAdapter, "dboToEntity").mockImplementation(jest.fn());
        });

        afterAll(() => {
            mockCursorFind.mockRestore();
        });

        it("should call cursorFind with provider filter", () => {
            paymentFlatPort.cursorFindChorusOnly();
            expect(mockCursorFind).toHaveBeenCalledWith({ provider: "chorus" });
        });

        it("should call cursorFind with provider and exerciceBudgetaire filter", () => {
            const exerciceBudgetaire = 2021;
            paymentFlatPort.cursorFindChorusOnly(exerciceBudgetaire);
            expect(mockCursorFind).toHaveBeenCalledWith({ provider: "chorus", exerciceBudgetaire: exerciceBudgetaire });
        });

    });
});
