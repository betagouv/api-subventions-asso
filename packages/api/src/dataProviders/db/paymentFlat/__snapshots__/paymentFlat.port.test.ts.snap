// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentFlat Port buildUpsertOperation returns bulk operation 1`] = `
{
  "updateOne": {
    "filter": {
      "uniqueId": "12345678901234-0001821732-2023-1-AC123-AC456-1680307200000-BRET-AA01/0776-C001-4000",
    },
    "update": {
      "$set": {
        "action": "Label d'action Exemple",
        "activite": "Label d'activité Exemple",
        "attachementComptable": "BRET",
        "codeAction": "AC123",
        "codeActivite": "AC456",
        "codeCentreFinancier": "AA01/0776-C001-4000",
        "dateOperation": 2023-04-01T00:00:00.000Z,
        "ej": "0001821732",
        "exerciceBudgetaire": 2023,
        "idEntrepriseBeneficiaire": "100000000",
        "idEtablissementBeneficiaire": "10000000000001",
        "idVersement": "12345678901234-0001821732-2023",
        "libelleCentreFinancier": "UO DGER XXXX-C001",
        "ministere": "Ministère Exemple",
        "mission": "Mission Exemple",
        "montant": 1000,
        "numeroProgramme": 1,
        "programme": "Programme Exemple",
        "provider": "chorus",
        "regionAttachementComptable": "Bretagne",
        "sigleMinistere": "ME",
        "typeIdEntrepriseBeneficiaire": "siren",
        "typeIdEtablissementBeneficiaire": "siret",
        "uniqueId": "12345678901234-0001821732-2023-1-AC123-AC456-1680307200000-BRET-AA01/0776-C001-4000",
      },
    },
    "upsert": true,
  },
}
`;
